### PR TITLE
chore(cli): add share-project-config-on/off to configure

### PR DESCRIPTION
**Pull request by 0618**
#### Description of changes

This PR is a part of the data consent re-work.
This PR 
- enables the user to toggle `shareProjectConfig` by runing `amplify configure share-project-config-on(off)`.
- update `amplify configure -h`
- add feedback once the user opt-in or opt-out
- fix typo `non sensitive` -> `non-sensitive`

|   | Before  | After  |
|---|---|---|
| configure -h  | image  | image  |
| init  | image  |  image image |

image

More upcoming changes for 
[ ] Show telemetry disclaimer (if not opted-out) on any first command which collecting telemetry data
[ ] Clean up old telemetry disclaimer in postinstall
[ ] Update docs

Ticket:
NA

#### Description of how you validated changes

Verified on local `amplify-dev`

#### Checklist

- [x] PR description included
- [ ] `yarn test` passes
- [ ] Tests are [changed or added](https://github.com/aws-amplify/amplify-cli/blob/dev/CONTRIBUTING.md#tests)
- [ ] Relevant documentation is changed or added (and PR referenced)
- [ ] New AWS SDK calls or CloudFormation actions have been added to relevant test and service IAM policies
- [ ] [Pull request labels](https://github.com/aws-amplify/amplify-cli/blob/dev/CONTRIBUTING.md#labels) are added

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by 0618** [comment]
CCI e2e: https://app.circleci.com/pipelines/github/aws-amplify/amplify-cli?branch=run-e2e/zhamujun/data-consent-configure

File: packages/amplify-cli/src/commands/configure.ts [hunk 1]
```diff
@@ -15,12 +16,25 @@ export const run = async (context: Context) => {
...
+    context.print.success('Share Project Config has been turned off');
+    return;
+  }
+
+  if (context.parameters.options['share-project-config-on']) {
+    DebugConfig.Instance.setAndWriteShareProject(true);
+    context.print.success('Share Project Config has been turned on');
+    return;
+  }
+
```
Comment (0618) [inline packages/amplify-cli/src/commands/configure.ts:37]: These lines are so spaghetti 😂  any suggestions how we can optimize it?

File: packages/amplify-cli/src/init-steps/s9-onSuccess.ts [hunk 2]
```diff
@@ -65,6 +65,13 @@ export const onSuccess = async (context: $TSContext): Promise<void> => {
 
     await FeatureFlags.ensureDefaultFeatureFlags(true);
     const result = await prompter.yesOrNo('Help improve Amplify CLI by sharing non sensitive configurations on failures', false);
```
Comment (josefaidt) [inline packages/amplify-cli/src/init-steps/s9-onSuccess.ts:67]:
```suggestion
    const result = await prompter.yesOrNo('Help improve Amplify CLI by sharing non-sensitive project configuration on failures', false);
```

File: packages/amplify-cli-core/src/help/commands-info.ts [hunk 3]
```diff
@@ -59,7 +59,12 @@ export const commandsInfo: Array<CommandInfo> = [
     command: 'configure',
     commandDescription: 'Configure the CLI to work with your AWS profile',
     commandUsage: 'amplify configure [subcommand]',
-    commandFlags: [],
+    commandFlags: [
+      { short: '', long: 'usage-data-on', flagDescription: 'Turn on usage data sharing.' },
+      { short: '', long: 'usage-data-off', flagDescription: 'Turn off usage data sharing.' },
+      { short: '', long: 'share-project-config-on', flagDescription: 'Turn on non-sensitive configurations sharing on failures.' },
```
Comment (josefaidt) [inline packages/amplify-cli-core/src/help/commands-info.ts:65]:
```suggestion
      { short: '', long: 'share-project-config-on', flagDescription: 'Turn on non-sensitive project configuration sharing on failures.' },
```

File: packages/amplify-cli-core/src/help/commands-info.ts [hunk 4]
```diff
@@ -59,7 +59,12 @@ export const commandsInfo: Array<CommandInfo> = [
...
+      { short: '', long: 'share-project-config-off', flagDescription: 'Turn off non-sensitive configurations sharing on failures.' },
```
Comment (josefaidt) [inline packages/amplify-cli-core/src/help/commands-info.ts:66]:
```suggestion
      { short: '', long: 'share-project-config-off', flagDescription: 'Turn off non-sensitive project configuration sharing on failures.' },
```

File: packages/amplify-cli-core/src/help/commands-info.ts (hunk 3 above)
Comment (josefaidt) [reply packages/amplify-cli-core/src/help/commands-info.ts:65]: adding "project configuration" to help bridge the understanding with its flag "share-project-config-on/off"

Comment (0618) [reply packages/amplify-cli-core/src/help/commands-info.ts:65]: Cool. I'll do a global search to update it everywhere.

File: packages/amplify-cli/src/commands/configure.ts (hunk 1 above)
Comment (sobolk) [reply packages/amplify-cli/src/commands/configure.ts:37]: at least it's not spaghetti on christmas tree like in other places ...

**Review by sobolk — DISMISSED** [review]
LGTM.
Feel free to merge when you get clean e2e run.

File: packages/amplify-cli/src/commands/configure.ts [hunk 5]
```diff
@@ -15,12 +16,25 @@ export const run = async (context: Context) => {
     context.print.success('Usage Data has been turned off');
     return;
   }
-  if (context.parameters.options['usage-data-on']) {
+
+  if (context.parameters.options['usage-data-on ']) {
```
Comment (jhockett) [inline packages/amplify-cli/src/commands/configure.ts:20]:
```suggestion
  if (context.parameters.options['usage-data-on']) {
```

File: packages/amplify-cli/src/commands/configure.ts [hunk 6]
```diff
@@ -15,12 +16,25 @@ export const run = async (context: Context) => {
...
     write(context, { usageDataConfig: { isUsageTrackingEnabled: true } });
     context.print.success('Usage Data has been turned on');
     return;
   }
 
+  if (context.parameters.options['share-project-config-off']) {
+    DebugConfig.Instance.setAndWriteShareProject(false);
+    context.print.success('Share Project Config has been turned off');
```
Comment (jhockett) [inline packages/amplify-cli/src/commands/configure.ts:28]: context.print operations are deprecated consider using `printer` from the `@aws-amplify/amplify-prompts` package, but not a merge blocker.

Comment (0618) [reply packages/amplify-cli/src/commands/configure.ts]: Done 3a7220a335dc8aa1e4243982f973137ddd616726

File: packages/amplify-e2e-core/src/utils/nexpect.ts [hunk 7]
```diff
@@ -753,7 +753,7 @@ export function nspawn(command: string | string[], params: string[] = [], option
   // For push operations in E2E we have to explicitly disable the Amplify Console App creation
   // as for the tests that need it, it is already enabled for init, setting the env var here
   // disables the post push check we have in the CLI.
-  if (params.length > 0 && params.find((param: string) => param.toLowerCase() === 'push')) {
+  if (params.length > 0 && params.some((param: string) => param.toLowerCase() === 'push')) {
```
Comment (0618) [inline packages/amplify-e2e-core/src/utils/nexpect.ts:756]: This change is not related to this PR...just came across it when reading the code

File: packages/amplify-e2e-core/src/configure/index.ts [hunk 8]
```diff
@@ -41,42 +50,48 @@ const authenticationOptions = ['AWS profile', 'AWS access keys'];
...
-
-  return new Promise((resolve, reject) => {
-    const chain = spawn(getCLIPath(), ['configure'], { stripColors: true })
-      .wait('Sign in to your AWS administrator account:')
-      .wait('Press Enter to continue')
-      .sendCarriageReturn()
-      .wait('Specify the AWS Region');
-
-    singleSelect(chain, s.region, amplifyRegions);
+export function amplifyConfigure(cwd: string | null, settings: AmplifyConfiguration | null, flag?: CommandFlags): Promise<void> {
```
Comment (sobolk) [inline packages/amplify-e2e-core/src/configure/index.ts:53]: If `settings` and `flag` are mutually exclusive (they are by looking at implementation) then please either change api to have union type (better) OR add validation to this function and throw if both are provided (not bad, but longer feedback loop than compiler error).

Yet other option is to split this function into two `amplifyConfigure` as is originally and `amplifyConfigureTelemetryFlags`.


```suggestion
export function amplifyConfigure(cwd: string | null, settings: AmplifyConfiguration | CommandFlags | null): Promise<void> {
```

File: packages/amplify-e2e-tests/src/__tests__/amplify-configure.test.ts [hunk 9]
```diff
@@ -53,6 +73,8 @@ function testAmplifyConfigureValidation(): Promise<void> {
       .wait('You must enter a valid secretAccessKey')
       .sendLine(validMockAWSSecretAccessKey)
       .wait('Profile Name:')
+      // 'config-test' is a new profile name using defaultAWSSecretAccessKey and validMockAWSSecretAccessKey.
+      // Do NOT use TEST_PROFILE_NAME, otherwise, it will override the default profile.
```
Comment (sobolk) [inline packages/amplify-e2e-tests/src/__tests__/amplify-configure.test.ts:77]:
```suggestion
      // Fake credentials used in this test are always invalid therefore
      // 'config-test' is a new profile name using defaultAWSSecretAccessKey and validMockAWSSecretAccessKey
      // to avoid overwriting profiles used in other tests.
      // Do NOT use TEST_PROFILE_NAME here, otherwise, it will override the default profile.
```

**Review by sobolk — COMMENTED** [review]
LGTM.

File: packages/amplify-e2e-core/src/configure/index.ts (hunk 8 above)
Comment (0618) [reply packages/amplify-e2e-core/src/configure/index.ts:53]: They are not really "mutually exclusive". Maybe at this moment, but there are several scenarios we may need both
- if we add tests for subcommands without creating a new function (e.g. `amplifyConfigureProject`)
- if we add a new flag with requires an option

So I think It's more scalable to separate them into 2 different variables.

Ideally, it'd be even better if we could use a destracture IMO

Comment (sobolk) [reply packages/amplify-e2e-core/src/configure/index.ts:53]: `settings` are not consumed if `flag` is passed at all. They are mutually exclusive effectively.
Besides scenarios in both logical branches in that if conditions are never going to be combined together. I.e. you won't ever set telemetry flags AND configure profiles.

```
if (flag) {
    return spawn(getCLIPath(), ['configure', `--${flag}`], { cwd, stripColors: true })
      .wait(commandFlagsReturnMessage[flag])
      .runAsync();
  } else {
    const allSettings = { ...defaultSettings, ...settings };
    const missingParam = MANDATORY_PARAMS.filter((p) => !Object.keys(allSettings).includes(p));
```

Comment (sobolk) [reply packages/amplify-e2e-core/src/configure/index.ts:53]: > So I think It's more scalable to separate them into 2 different variables.

Define `scalable`

Comment (sobolk) [reply packages/amplify-e2e-core/src/configure/index.ts:53]: This is classic SRP problem.

Comment (0618) [reply packages/amplify-e2e-core/src/configure/index.ts]: I see. That's fair. Done 049cdcd7cec03d1d33f979861284b2884258417b

**Review by sobolk — APPROVED** [review]

**Review by joebuono — APPROVED** [review]
